src/webapp/components/search-input/SearchInput.tsx

```diff
@@ -0,0 +1,115 @@
+import React, { useCallback, useEffect, useState } from "react";
+import { TextField } from "@material-ui/core";
+import { IconSearch24 } from "@dhis2/ui";
+import styled from "styled-components";
+
+import i18n from "../../../utils/i18n";
+
+type SearchInputProps = {
+    value: string;
+    onChange: (event: string) => void;
+    placeholder?: string;
+    disabled?: boolean;
+};
+
+export const SearchInput: React.FC<SearchInputProps> = React.memo(
+    ({ value, onChange, placeholder = "", disabled = false }) => {
+        const [stateValue, updateStateValue] = useState(value);
+
+        useEffect(() => updateStateValue(value), [value]);
+
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        const onChangeDebounced = useCallback(
+            debounce((value: string) => {
+                if (onChange) onChange(value);
+            }, 400),
+            [onChange]
+        );
+
+        const handleChange = useCallback(
+            (event: React.ChangeEvent<HTMLInputElement>) => {
+                const value = event.target.value;
+                onChangeDebounced(value);
+                updateStateValue(value);
+            },
+            [onChangeDebounced, updateStateValue]
+        );
+
+        const handleKeydown = useCallback((event: React.KeyboardEvent<HTMLInputElement>) => {
+            event.stopPropagation();
+        }, []);
+
+        return (
+            <Container>
+                <IconContainer $disabled={disabled}>
+                    <IconSearch24 />
+                </IconContainer>
+                <StyledTextField
+                    onChange={handleChange}
+                    placeholder={placeholder || i18n.t("Search")}
+                    value={stateValue}
+                    role="searchbox"
+                    onKeyDown={handleKeydown}
+                    disabled={disabled}
+                    variant="outlined"
+                    inputProps={{ "aria-label": "search" }}
+                />
+            </Container>
+        );
+    }
+);
+
+function debounce<F extends (...args: any[]) => any>(func: F, delay: number) {
+    let timeout: ReturnType<typeof setTimeout> | null = null;
+
+    const debounced = (...args: Parameters<F>): void => {
+        if (timeout !== null) {
+            clearTimeout(timeout);
+        }
+        timeout = setTimeout(() => func(...args), delay);
+    };
+
+    return debounced as (...args: Parameters<F>) => ReturnType<F>;
```

Comment:
That looks wrong, debouced never returns anything, but the casting says it returns ` ReturnType<F>`.